common/task/priorityTaskProcessor.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package task
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+)
+
+type (
+	// PriorityTaskProcessorOptions configs PriorityTaskProcessor
+	PriorityTaskProcessorOptions struct {
+		WorkerCount int
+		RetryPolicy backoff.RetryPolicy
+	}
+
+	priorityTaskProcessorImpl struct {
+		scheduler PriorityScheduler
+
+		status       int32
+		workerWG     sync.WaitGroup
+		logger       log.Logger
+		metricsScope metrics.Scope
+		options      *PriorityTaskProcessorOptions
+	}
+)
+
+// NewPriorityTaskProcessor creates a new PriorityTaskProcessor
+func NewPriorityTaskProcessor(
+	scheduler PriorityScheduler,
+	logger log.Logger,
+	metricsClient metrics.Client,
+	options *PriorityTaskProcessorOptions,
+) PriorityTaskProcessor {
+	return &priorityTaskProcessorImpl{
```

Comment:
This relationship looks backwards from a mental model perspective i.e. scheduler should own a workerPool to schedule things and not the other way around. How about making workerPool (i.e. your taskProcessor) completely independent i.e. Something like the following:

```
Scheduler interface {
   Submit(Task task)
}

Task interface {
    ..
    Priority() int
}

TaskWorkerPool(poolSize, taskQ, metrics, logger)
 -- where taskQ is a small buffered input queue that feeds in tasks to be processed

```

What do you think ?

Reply:
Synced offline. Will implement priority task processor on top of parallel task processor (this pr) so that the logic for worker pool and processing tasks independent and can be reused.

Will send out another pr for the priority task processor (scheduler).